cmd/kubeadm/app/cmd/phases/join/kubelet.go

```diff
@@ -128,6 +130,39 @@ func runKubeletStartJoinPhase(c workflow.RunData) (returnErr error) {
 		return errors.Errorf("couldn't create client from kubeconfig file %q", bootstrapKubeConfigFile)
 	}
 
+	// Make sure to exit before TLS bootstrap if a node with the same name exist in the cluster.
+	// A new node with the same name as an existing control-plane node can cause undefined
+	// behavior and ultimatelly control-plane failure.
+	//
+	// Given there are multiple ways to define the node name we apply the following precedence
+	// from lower to higher:
+	// - actual hostname
+	// - JoinConfiguration.NodeRegistration.Name (same as "--node-name" passed to "kubeadm join")
+	// - "hostname-overide" flag in JoinConfiguration.NodeRegistration.KubeletExtraArgs
+	//
+	// The same precedence already applies in app/phases/kubelet/flags.go
+	nodeName, err := kubeadmutil.GetHostname("")
+	if err != nil {
+		return err
+	}
+	if cfg.NodeRegistration.Name != "" {
+		nodeName = cfg.NodeRegistration.Name
+	}
+	if name, ok := cfg.NodeRegistration.KubeletExtraArgs["hostname-override"]; ok {
+		nodeName = name
+	}
+
+	klog.V(1).Infof("[kubelet-start] Checking for an existing node with name %q in the cluster", nodeName)
+	nodes, err := bootstrapClient.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "cannot obtain the list of nodes in the cluster")
```

Comment:
should this be a skippable preflight check instead?

Reply:
no so sure about this, stale nodes are a subject of the administrator calling `kubectl delete`. `kubeadm join` failing during boostrap would indicate that the cluster requires admin intervetion.

if the user is allowed to skip this check, it can lead to catastrophic outcome - e.g. imaging a case where the operator responsible for the worker join is not the cluster admin.

let's talk more about this during the office hours, possibly next week.